watchtower/wtmock/client_db.go

```diff
@@ -67,8 +79,83 @@ func (m *ClientDB) CreateTower(lnAddr *lnwire.NetAddress) (*wtdb.Tower, error) {
 	return copyTower(tower), nil
 }
 
-// LoadTower retrieves a tower by its tower ID.
-func (m *ClientDB) LoadTower(towerID wtdb.TowerID) (*wtdb.Tower, error) {
+// RemoveTower modifies a tower's record within the database. If an address is
+// provided, then _only_ the address record should be removed from the tower's
+// persisted state. Otherwise, we'll attempt to mark the tower as inactive by
+// marking all of its sessions inactive. If any of its sessions has unacked
+// updates, then ErrTowerUnackedUpdates is returned. If the tower doesn't have
+// any sessions at all, it'll be completely removed from the database.
+//
+// NOTE: An error is not returned if the tower doesn't exist.
+func (m *ClientDB) RemoveTower(pubKey *btcec.PublicKey, addr net.Addr) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	tower, err := m.loadTower(pubKey)
+	if err == wtdb.ErrTowerNotFound {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if addr != nil {
+		tower.RemoveAddress(addr)
+		m.towers[tower.ID] = tower
+		return nil
+	}
+
+	towerSessions, err := m.listClientSessions(&tower.ID)
+	if err != nil {
+		return err
+	}
+	if len(towerSessions) == 0 {
+		var towerPK towerPK
+		copy(towerPK[:], pubKey.SerializeCompressed())
+		delete(m.towerIndex, towerPK)
+		delete(m.towers, tower.ID)
+		return nil
+	}
+
+	for id, session := range towerSessions {
+		if len(session.CommittedUpdates) > 0 {
+			return wtdb.ErrTowerUnackedUpdates
+		}
+		session.Status = wtdb.CSessionInactive
+		m.activeSessions[id] = session
+	}
+
+	return nil
+}
+
+// LoadTower retrieves a tower by its public key.
+func (m *ClientDB) LoadTower(pubKey *btcec.PublicKey) (*wtdb.Tower, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.loadTower(pubKey)
+}
+
+// loadTower retrieves a tower by its public key.
+//
+// NOTE: This method requires the database's lock to be acquired.
+func (m *ClientDB) loadTower(pubKey *btcec.PublicKey) (*wtdb.Tower, error) {
+	var towerPK towerPK
+	copy(towerPK[:], pubKey.SerializeCompressed())
+
+	towerID, ok := m.towerIndex[towerPK]
```

Comment:
Out of the scope of this PR, so just reflecting, but it seems that we can remove much of the code in this mock tower, if we abstracted the storage backend from the actual tower DB a bit. So in this case, the business logic would be shared across the regular and the mock, with the only difference being storage in a map vs actually on disk. 

Each time we modify the behavior of the _actual_ tower DB, we must also take care to ensure that this is also updated, and implements the actual behavior perfectly (ideally). This adds an additional cost/diff to modifying this area of the codebase, as we essentially need two maintain the logic of two structs that do the same thing.

Reply:
The mock DB had it place during initial development as we wanted to flesh out the biz logic without being burdened with actual DB details such as bucket structure, serialization, etc. However at this point, I don't see a good reason for it to exists in its current state.

Reply:
In other parts of the codebase we have mock DBs for unit tests to speed them up, but doesn't seem like there's something to gain here given that the tests still run with the actual bolt implementation.

Reply:
the db tests run both mock and boltdb implementations to assert they have identical behavior. all of the other packages use the mockdb to speed them up